File: app/src/main/java/tw/tonyyang/englishwords/WordListM1Fragment.java
package tw.tonyyang.englishwords;

import android.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import org.androidannotations.annotations.AfterViews;
import org.androidannotations.annotations.EFragment;
import org.androidannotations.annotations.ViewById;
import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;


@EFragment(R.layout.fragment_word_list)
public class WordListM1Fragment extends Fragment {
    private static final Logger logger = LoggerFactory.getLogger(WordListM1Fragment.class);

    @ViewById(R.id.recyclerview_word_list)
    RecyclerView recyclerView;

    private WordListM1Adapter wordListM1Adapter;

    @AfterViews
    protected void initViews() {
        recyclerView.setHasFixedSize(true);
        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(getActivity());
        recyclerView.setLayoutManager(layoutManager);
        wordListM1Adapter = new WordListM1Adapter(getAllCategory());
        wordListM1Adapter.setOnRecyclerViewListener(onRecyclerViewListener);
        recyclerView.setAdapter(wordListM1Adapter);
    }

    private List<String> getAllCategory() {
        return App_.getDb().userDao().getAllCategory();
    }

    private WordListM1Adapter.OnRecyclerViewListener onRecyclerViewListener = new WordListM1Adapter.OnRecyclerViewListener() {

        @Override
        public void onItemClick(View v, int position) {
            String category = wordListM1Adapter.getItem(position);
            WordListM2Activity_.intent(getActivity())
                    .category(category)
                    .start();
        }

        @Override
        public void onItemLongClick(View v, int position) {

        }
    };

    @Override
    public void onStart() {
        super.onStart();
        if (!EventBus.getDefault().isRegistered(this)) {
            EventBus.getDefault().register(this);
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (EventBus.getDefault().isRegistered(this)) {
            EventBus.getDefault().unregister(this);
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onRealTimeUpdateEvent(RealTimeUpdateEvent event) {
        RealTimeUpdateEvent.Type type = event.getType();
        if (type == RealTimeUpdateEvent.Type.UPDATE_WORD_LIST) {
            logger.debug("UPDATE_WORD_LIST: " + event.getMessage());
            wordListM1Adapter.setWordList(getAllCategory());
            wordListM1Adapter.notifyDataSetChanged();
        }
    }
}

File: app/src/main/java/tw/tonyyang/englishwords/WordListM1Fragment.kt
package tw.tonyyang.englishwords

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import kotlinx.coroutines.*
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import tw.tonyyang.englishwords.App.Companion.db
import tw.tonyyang.englishwords.databinding.FragmentWordListBinding

class WordListM1Fragment : Fragment() {

    private val wordListM1Adapter: WordListM1Adapter by lazy {
        WordListM1Adapter()
    }

    private lateinit var binding: FragmentWordListBinding

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        binding = FragmentWordListBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        wordListM1Adapter.onRecyclerViewListener = onRecyclerViewListener
        binding.recyclerview.apply {
            setHasFixedSize(true)
            layoutManager = LinearLayoutManager(activity)
            adapter = wordListM1Adapter
        }
        updateCategoryList()
    }

    private fun updateCategoryList() = lifecycleScope.launch(Dispatchers.Main) {
        withContext(Dispatchers.Default) {
            db.userDao().allCategory
        }.let {
            wordListM1Adapter.categoryList = it
        }
    }

    private val onRecyclerViewListener: WordListM1Adapter.OnRecyclerViewListener = object : WordListM1Adapter.OnRecyclerViewListener {
        override fun onItemClick(v: View?, position: Int) {
            val category = wordListM1Adapter.getItem(position)
            val intent = Intent(activity, WordListM2Activity::class.java)
            val bundle = Bundle()
            bundle.putString(WordListM2Activity.EXTRA_CATEGORY, category)
            intent.putExtras(bundle)
            startActivity(intent)
        }

        override fun onItemLongClick(v: View?, position: Int) {
            // do nothing
        }
    }

    override fun onStart() {
        super.onStart()
        if (!EventBus.getDefault().isRegistered(this)) {
            EventBus.getDefault().register(this)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        if (EventBus.getDefault().isRegistered(this)) {
            EventBus.getDefault().unregister(this)
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onRealTimeUpdateEvent(event: RealTimeUpdateEvent) {
        val type = event.type
        if (type === RealTimeUpdateEvent.Type.UPDATE_WORD_LIST) {
            Logger.d(TAG, "UPDATE_WORD_LIST: " + event.message)
            updateCategoryList()
        }
    }

    companion object {
        private val TAG = WordListM1Fragment::class.java.simpleName
    }
}